---
title: "Species redistribution"
author: "Brunno F Oliveira"
date: "9/28/2021"
output:   
  html_document:
    toc: true
    toc_float: true
    code_folding: hide 
---

Fix predictors and occurrence data

# Setup
```{r packages, message=FALSE, include=FALSE}

rm(list=ls())
gc()

list.of.packages <- c("raster","spdep","rgdal","maptools","sp","spThin","sf",
                      "gridExtra","viridis","ggplot2","cowplot",
                      "reshape2","tidyr","tidyverse","data.table",
                      "RCurl","curl","sqldf","RSQLite",
                      "foreach","doParallel","pbapply",
                      "V.PhyloMaker","phytools",
                      "TNRS","httr","jsonlite")

new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]

if(length(new.packages)) install.packages(new.packages)

sapply(list.of.packages, require, character.only = TRUE)

```

# create an empty map
```{r empty}

# create an empty raster
empty.r <- raster(xmn=-180, xmx=180, ymn=-60, ymx=90)
res(empty.r) <- 0.08333 # 10 km
# res(empty.r) <- 0.041665 # 5 km
# res(empty.r) <- 0.008333 # 1 km
crs(empty.r) <- crs("+proj=longlat +datum=WGS84 +no_defs")

```

# Load occurrences
```{r load}

occur.set <- list.files("D:/GIS/Biodiversity/Distributions/BIEN/occur",full.names = T)

# There was no obs in some chunks because all obs were 1) invasive, or 2) cultivar observations, or 3) ?

# delete these chunks
tokeep <- pbsapply(occur.set, function(i) {
  tmp <- readRDS(i)
  dim(tmp)[1]
})
summary(tokeep)
torem <- which(tokeep==0)
# new occur set
occur.set <- occur.set[-torem]

#################
# Total N species
sp.list <- pbsapply(occur.set, function(x){
  tmp <- readRDS(x)
  unique(tmp$scrubbed_species_binomial)
})
sp.list <- unlist(sp.list)

# N species
length(sp.list) # 175365

#################
# Total N species >=10 obs
sp.list10 <- pbsapply(occur.set, function(x){
  tmp <- readRDS(x)
  tmp <- tmp$scrubbed_species_binomial
  tmp[which(table(tmp)>=10)]
})
sp.list10 <- unlist(sp.list10)

# N species >=10 obs
length(sp.list10) # 100583

```


# Prepare data
```{r prep}

# Thin occurrences
# Get only sps with >=10 obs
cl <- makeCluster(detectCores())
clusterExport(cl, c("empty.r","occur.set","crs","rbindlist"))

occs <- pblapply(occur.set, function(i) {
  tmp <- readRDS(i)
  tmp <- tmp[,1:3] # read only sp names and coords
  
  # Select species with >=10 obs
  sptogo <- table(tmp$scrubbed_species_binomial)
  sptogo <- names(sptogo)[sptogo>=10]
  tmp <- tmp[which(tmp$scrubbed_species_binomial %in% sptogo),]
  # Thin points to remove spatial autocorrelation
  # get only one obs per cell for each species
  sptogo <- unique(tmp$scrubbed_species_binomial)
  if(length(sptogo)>0){
    occs.thinned <- list()
    for(x in 1:length(sptogo)) { cat("\r",x,"from",length(sptogo))
      data.tmp <- tmp[which(tmp$scrubbed_species_binomial %in% sptogo[x]),]
      sp::coordinates(data.tmp)<- ~ longitude + latitude 
      sp::proj4string(data.tmp)<- crs(empty.r)
      # Count of observations
      # get the bounding box
      bb <- sp::bbox(data.tmp)
      # make extent object and buffer by 1 degree
      bb <- raster::extent(bb[1]-1, bb[3]+1, bb[2]-1, bb[4]+1)
      # crop environmental rasters by the extent to make the background extent
      bb <- raster::crop(empty.r, bb)
      
      data.tmp <- raster::rasterize(data.tmp, bb, 
                                    field="scrubbed_species_binomial",
                                    fun="count")
      # From raster to data.frame
      data.tmp <- na.omit(raster::as.data.frame(data.tmp, xy = T))
      data.tmp[,3] <- sptogo[x]
      names(data.tmp) <- c("x","y","sps")
      occs.thinned[[x]] <- data.tmp
    }
    
    occs.thinned <- rbindlist(occs.thinned)
    
    # Select species with >=10 observations after thinned
    sptogo <- table(occs.thinned$sps)
    sptogo <- names(sptogo)[sptogo>9]
    occs.thinned <- occs.thinned[which(occs.thinned$sps %in% sptogo),]
    
    # write set to disk
    # only save if data size is >=10 (in case only one species in the data frame)
    if(nrow(occs.thinned)>=10){
      if(!dir.exists("C:/BIEN_occur_thinned")){
      dir.create("C:/BIEN_occur_thinned")
    }
    var2save <- strsplit(i,"/occur/")[[1]][2]
    saveRDS(occs.thinned,
            paste0("C:/BIEN_occur_thinned/", var2save))
    }
  }
},
cl = cl) 

stopCluster(cl)
# takes 9 minutes in 10km ~ 0.08333 Dg
# takes 35 minutes in 5km ~ 0.041665 Dg
# takes x minutes in 1km ~ 0.008333 Dg

new.occur.set <- list.files("C:/BIEN_occur_thinned/",
                            full.names = T)

##########################
# test if min n species in a chunk is 10
# 
# N_sps <- pbsapply(new.occur.set, function(x){
#   tmp <- readRDS(x)
#   nrow(tmp)
# })
# summary(N_sps)

##########################
# count new N species
new.sp.list <- pbsapply(new.occur.set, function(x){
  tmp <- readRDS(x)
  unique(tmp$sps)
})
new.sp.list <- unlist(new.sp.list)

# N species
length(new.sp.list)
# 80338 species with resolution of 10km ~ 0.08333 Dg
# 82655 species with resolution of 5km ~ 0.041665 Dg



```

## Resolve species names using TNRS
Resolve names based on the TAXONOMIC NAME RESOLUTION SERVICE V5.0
```{r res_sps, eval=FALSE, message=FALSE, warning=FALSE}

# all names
Ori = new.sp.list
length(unique(Ori)) # 80338

# Remove plant names containing 'sp.', 'subsp.', 'subgen.', 'sect.' and 'cf.'
if(any(grep('sp\\.',Ori))){ Ori <- Ori[-grep('sp\\.',Ori)] }
if(any(grep('subgen\\.',Ori))){ Ori <- Ori[-grep('subgen\\.',Ori)] }
if(any(grep('sect\\.',Ori))){ Ori <- Ori[-grep('sect\\.',Ori)] }
if(any(grep('cf\\.',Ori))){ Ori <- Ori[-grep('cf\\.',Ori)]  }

length(unique(Ori)) # 80338

# Run in parallel for faster results
spchunks <- split(Ori, sort(rep_len(1:160, length(Ori))))
# N species in each chunck
# head(lapply(spchunks, length))

cl <- makeCluster(detectCores())
clusterExport(cl, c("spchunks", "TNRS"))

TNRS_names <- pblapply(spchunks, function(i) {
  test <- TNRS(taxonomic_names = i)
  while(length(test)==0){ # prevent error
    test <- TNRS(taxonomic_names = i)
  }
  return(test)
  }
  , cl = cl
)

stopCluster(cl)
# took 07m 30s

head(lapply(TNRS_names, dim))

# check if any chunk fail
cfail <- sapply(TNRS_names, function(x) {
  if(length(x)==0){ 1 }
  else{ 0 }
})
# N chunks that fail
length(which(cfail==1))

TNRS_names <- do.call("rbind", TNRS_names)

# remove species not found
TNRS_names <- TNRS_names[-which(TNRS_names$Overall_score==""),]

# remove bryophytes
taxonomy <- BIEN::BIEN_taxonomy_species(gsub("_"," ",TNRS_names$Name_matched))
length(unique(taxonomy$scrubbed_species_binomial))
length(unique(TNRS_names$Name_matched))
taxonomy2 <- taxonomy[-which(duplicated(taxonomy$scrubbed_species_binomial)),]
TNRS_names_ <- merge(taxonomy2[,c("scrubbed_species_binomial","higher_plant_group")], TNRS_names, 
                     by.x = "scrubbed_species_binomial", by.y = "Name_matched",
                     all.x = T)
# remove bryophytes
TNRS_names_ <- TNRS_names_[-which(TNRS_names_$higher_plant_group=="bryophytes"),]

my_sp_list <- unique(TNRS_names_$scrubbed_species_binomial)
length(my_sp_list)
# N species = 79670

taxonomy <- TNRS_names_[,c(1,2,13,15)]
taxonomy <- taxonomy[which(taxonomy$scrubbed_species_binomial%in%my_sp_list),]
taxonomy <- taxonomy[-which(duplicated(taxonomy$scrubbed_species_binomial)),]
names(taxonomy) <- c("sps","Higher_Group","Family","Genus")

```

# Create phylogeny for plants
```{r}
# load phylogeny from Jin & Qian 2019
tree <- V.PhyloMaker::GBOTB.extended
tree$tip.label <- gsub("_"," ", tree$tip.label)
sp.tree <- tree$tip.label
# fix sp.tree names
# Run in parallel for faster results
Ori <- sp.tree
spchunks <- split(Ori, sort(rep_len(1:160, length(Ori))))
# N species in each chunck
# head(lapply(spchunks, length))

cl <- makeCluster(detectCores())
clusterExport(cl, c("spchunks", "TNRS"))

tree_names <- pblapply(spchunks, function(i) {
  test <- TNRS(taxonomic_names = i)
  while(length(test)==0){ # prevent error
    test <- TNRS(taxonomic_names = i)
  }
  return(test)
}
, cl = cl
)

stopCluster(cl)
# took 07m 16s

head(lapply(tree_names, dim))

# check if any chunk fail
cfail <- sapply(tree_names, function(x) {
  if(length(x)==0){ 1 }
  else{ 0 }
})
# N chunks that fail
length(which(cfail==1))

tree_names <- do.call("rbind", tree_names)

# fix names
tree$tip.label <- tree_names$Name_matched
length(which(duplicated(tree$tip.label)))
# remove duplicated branches
tree <- drop.tip(tree, which(duplicated(tree$tip.label)))
# remove species not found
tree <- drop.tip(tree, which(tree$tip.label=="[No match found]"))
# remove species not at the species list
tree <- keep.tip(tree, which(tree$tip.label%in%my_sp_list))
# sps in tree
sp.tree <- tree$tip.label
# taxonomy of species in tree
treetax <- taxonomy[which(taxonomy$sps %in% sp.tree),]
treetax <- data.frame(species = treetax$sps, Genus = treetax$Genus, Family = treetax$Family)
# Extract the genus- and family-level nodes information in a phylogeny (necessary for using user defined tree in phylo.maker) 
treenodes <- build.nodes.1(tree,
                           treetax)
# species to be added to the tree
sp2add <- data.frame(species = taxonomy$sps, 
                     Genus = taxonomy$Genus, 
                     Family = taxonomy$Family)
# Create phylogeny with all species
start_time <- Sys.time()
tree <- phylo.maker(sp.list = sp2add, 
                    tree = tree,
                    nodes = treenodes)

end_time <- Sys.time()
end_time - start_time
# took 38 min

# length species fail to bind
failtobind <- tree$species.list
table(failtobind$status) # 109
tree = tree$scenario.3

# keep same species in tree and taxonomy
spincommon <- intersect(gsub("_"," ",tree$tip.label), taxonomy$sps)

taxonomy <- taxonomy[which(taxonomy$sps %in% spincommon),]
saveRDS(taxonomy, "Data/plants_taxonomy.RDS")

tree <- keep.tip(tree, gsub(" ","_",spincommon))
saveRDS(tree,"Data/plants_phy.RDS")

my_sp_list <- spincommon
length(my_sp_list)
# N species = 79556

```

# Phylogenetic eigen veectors
```{r}
# It will be useful for trait imputattion
# add phylogenetic eigen vectors (at the genus level)
tips<-as.character(tree$tip.label)
genera<-sapply(strsplit(tips,"_"),function(x) x[1]) 
spgen <- data.frame(tips, genera)
genera.list <- unique(genera)
# randomly choose one species per the genera to account for uncertainty across phylogenies
ii <- tips[-which(duplicated(genera))]
# get the genera level tree
treeg <- drop.tip(tree, setdiff(tree$tip.label, ii))
# calculate eigenvectors for each tree
x <- PVRdecomp(treeg)
# Get eigenvectors
eigs <- x@Eigen$values
# Calc proportion explained by eigenvectors
rbind(
  SD = sqrt(eigs),
  Proportion = eigs/sum(eigs),
  Cumulative = cumsum(eigs)/sum(eigs))[,1:30]
 
eigssumres <- data.frame(t(rbind(
  SD = sqrt(eigs),
  Proportion = eigs/sum(eigs),
  Cumulative = cumsum(eigs)/sum(eigs))))

write.csv(eigssumres, "Results/plants_eigenPropExp.csv")

# Cumulative prop explained by eigens
# plot((cumsum(eigs)/sum(eigs))[1:30])

# select the first 30 eigenvectors, explaining 53% of the variation of the whole phylogeny, thus avoiding including low-representative axes in the analysis (>1%).
vecs <- x@Eigen$vectors

# create table of eigens per species >> from genus level to species level
gen.vecs <- data.frame(gen = genera.list, vecs)
sp.vecs <- merge(data.frame(sps = tips, gen = genera), gen.vecs, by = "gen")[,-1]
sp.vecs$sps <- gsub("_"," ",sp.vecs$sps)

# save eigen vectors
saveRDS(sp.vecs, "Data/plants_PhyloEigenVect.RDS")

```


# Create a SQL file with species occurrences
```{r}

my_db_file <- "R/3_SDM/Cluster/plant_occur.sqlite"
my_db <- src_sqlite(my_db_file, create = TRUE)

# driver
drv <- dbDriver("SQLite") 
# connect
db <- dbConnect(drv, my_db_file) 

# add occurrences to the DB
pbsapply(new.occur.set, function(x){
  tmp <- readRDS(x)
  # fix sp names
  tmp <- merge(tmp, TNRS_names_[,c("scrubbed_species_binomial","Name_submitted")], 
               by.x = "sps", by.y = "Name_submitted",
               all.x = T)
  tmp <- tmp[,2:4]
  names(tmp) <- c("x","y","sps")
  # get only species in phylogeny
  if(any(tmp$sps %in% my_sp_list)){
    tmp <- tmp[which(tmp$sps %in% my_sp_list),]
    dbWriteTable(conn=db, name="occur", tmp, append=T, row.names=F)
  }
})

# disconnect
DBI::dbDisconnect(db)

```

# test a query
```{r}

# driver
drv <- dbDriver("SQLite") 
# connect
db <- dbConnect(drv, my_db_file) 

# get all sps
x1 = dbGetQuery(db, "
   SELECT * FROM occur
")

# get 100 sps
x2 = dbGetQuery(
  db, 
  paste(
    "SELECT *",
    "FROM occur",
    "WHERE sps in (", paste(shQuote(my_sp_list[1:100], type = "sh"),collapse = ', '), ")")
)

# disconnect
dbDisconnect(db)

```

# count N obs per species
```{r}

n_obs <- table(x1$sps)
n_obs <- as.data.frame(n_obs)
names(n_obs) <- c("Species", "N_obs")
# subset for my list (i.e., without bryophytes)
n_obs <- n_obs[n_obs$Species %in% my_sp_list,]

write.csv(n_obs, "R/3_SDM/plants_N_obs_sps.csv")

```

# save full species list
```{r}
saveRDS(my_sp_list, "R/3_SDM/Cluster/plants_list.RDS")
```

